File: patterns/Creational/Prototype/PHP/RealWorldExamples/Shape.php
<?php

/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Shape Example
|--------------------------------------------------------------------------
| Implement the Prototype Design Pattern to clone shape objects, 
| avoiding the need for subclass-specific object creation processes.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/milendenev0912/design-patterns
|--------------------------------------------------------------------------
*/

namespace Creational\Prototype\Example;

/**
 * Prototype Interface
 */
interface ShapePrototype
{
    public function __clone();
}

/**
 * Abstract Shape Class
 * Defines common properties and the cloning interface.
 */
abstract class Shape implements ShapePrototype
{
    public $color;

    public function setColor(string $color): void
    {
        $this->color = $color;
    }

    abstract public function draw(): string;

    /**
     * This method is required by the prototype pattern.
     */
    abstract public function __clone();
}

/**
 * Concrete Prototype: Circle
 */
class Circle extends Shape
{
    private $radius;

    public function __construct(int $radius)
    {
        $this->radius = $radius;
    }

    public function draw(): string
    {
        return "Drawing a circle with radius " . $this->radius . " and color " . $this->color;
    }

    public function __clone()
    {
        // Optionally customize the behavior of cloned objects.
    }
}

/**
 * Concrete Prototype: Rectangle
 */
class Rectangle extends Shape
{
    private $width;
    private $height;

    public function __construct(int $width, int $height)
    {
        $this->width = $width;
        $this->height = $height;
    }

    public function draw(): string
    {
        return "Drawing a rectangle with width " . $this->width . ", height " . $this->height . " and color " . $this->color;
    }

    public function __clone()
    {
        // Optionally customize the behavior of cloned objects.
    }
}

/**
 * Client Code demonstrating Prototype Design Pattern.
 */
function clientCode()
{
    // Create a Circle object.
    $circle = new Circle(10);
    $circle->setColor("Red");

    // Clone the Circle object.
    $clonedCircle = clone $circle;
    $clonedCircle->setColor("Blue");

    // Create a Rectangle object.
    $rectangle = new Rectangle(20, 10);
    $rectangle->setColor("Green");

    // Clone the Rectangle object.
    $clonedRectangle = clone $rectangle;
    $clonedRectangle->setColor("Yellow");

    // Output original and cloned shapes.
    echo "Original Circle: " . $circle->draw() . "\n";
    echo "Cloned Circle: " . $clonedCircle->draw() . "\n";

    echo "\nOriginal Rectangle: " . $rectangle->draw() . "\n";
    echo "Cloned Rectangle: " . $clonedRectangle->draw() . "\n";
}

clientCode();

File: patterns/Creational/Prototype/Go/RealWorldExamples/Shape.go
/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Shape Example
|--------------------------------------------------------------------------
| Implement the Prototype Design Pattern to clone shape objects, 
| avoiding the need for subclass-specific object creation processes.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/Milen Denev/design-patterns-in-go
|--------------------------------------------------------------------------
*/

package main

import (
    "fmt"
)

// ShapePrototype interface defines the clone method.
type ShapePrototype interface {
    Clone() ShapePrototype
    Draw() string
    SetColor(color string)
}

// Shape struct defines common properties and the cloning interface.
type Shape struct {
    Color string
}

// SetColor sets the color of the shape.
func (s *Shape) SetColor(color string) {
    s.Color = color
}

// Circle struct represents a concrete prototype.
type Circle struct {
    Shape
    Radius int
}

// NewCircle is the constructor for Circle.
func NewCircle(radius int) *Circle {
    return &Circle{Radius: radius}
}

// Draw outputs the properties of the circle.
func (c *Circle) Draw() string {
    return fmt.Sprintf("Drawing a circle with radius %d and color %s", c.Radius, c.Color)
}

// Clone creates a copy of the circle.
func (c *Circle) Clone() ShapePrototype {
    clone := *c
    return &clone
}

// Rectangle struct represents a concrete prototype.
type Rectangle struct {
    Shape
    Width  int
    Height int
}

// NewRectangle is the constructor for Rectangle.
func NewRectangle(width, height int) *Rectangle {
    return &Rectangle{Width: width, Height: height}
}

// Draw outputs the properties of the rectangle.
func (r *Rectangle) Draw() string {
    return fmt.Sprintf("Drawing a rectangle with width %d, height %d and color %s", r.Width, r.Height, r.Color)
}

// Clone creates a copy of the rectangle.
func (r *Rectangle) Clone() ShapePrototype {
    clone := *r
    return &clone
}

// Client code demonstrating Prototype Design Pattern.
func clientCode() {
    // Create a Circle object.
    circle := NewCircle(10)
    circle.SetColor("Red")

    // Clone the Circle object.
    clonedCircle := circle.Clone().(*Circle)
    clonedCircle.SetColor("Blue")

    // Create a Rectangle object.
    rectangle := NewRectangle(20, 10)
    rectangle.SetColor("Green")

    // Clone the Rectangle object.
    clonedRectangle := rectangle.Clone().(*Rectangle)
    clonedRectangle.SetColor("Yellow")

    // Output original and cloned shapes.
    fmt.Println("Original Circle: " + circle.Draw())
    fmt.Println("Cloned Circle: " + clonedCircle.Draw())

    fmt.Println("\nOriginal Rectangle: " + rectangle.Draw())
    fmt.Println("Cloned Rectangle: " + clonedRectangle.Draw())
}

func main() {
    clientCode()
}
